src/sync/sync_config.hpp

```diff
@@ -122,27 +131,28 @@ struct SyncConfig {
     // Set this field, but read from `realm_url()`.
     std::string reference_realm_url;
     SyncSessionStopPolicy stop_policy = SyncSessionStopPolicy::AfterChangesUploaded;
-    std::function<SyncBindSessionHandler> bind_session_handler = nullptr;
-    std::function<SyncSessionErrorHandler> error_handler = nullptr;
-    std::shared_ptr<ChangesetTransformer> transformer = nullptr;
-    util::Optional<std::array<char, 64>> realm_encryption_key = none;
+    std::function<SyncBindSessionHandler> bind_session_handler;
+    std::function<SyncSessionErrorHandler> error_handler;
+    std::shared_ptr<ChangesetTransformer> transformer;
+    util::Optional<std::array<char, 64>> realm_encryption_key;
     bool client_validate_ssl = true;
-    util::Optional<std::string> ssl_trust_certificate_path = none;
-    std::function<sync::Session::SSLVerifyCallback> ssl_verify_callback = nullptr;
-    util::Optional<ProxyConfig> proxy_config = none;
+    util::Optional<std::string> ssl_trust_certificate_path;
+    std::function<sync::Session::SSLVerifyCallback> ssl_verify_callback;
+    util::Optional<ProxyConfig> proxy_config;
     bool is_partial = false;
-    util::Optional<std::string> custom_partial_sync_identifier = none;
+    util::Optional<std::string> custom_partial_sync_identifier;
 
     bool validate_sync_history = true;
 
-    util::Optional<std::string> authorization_header_name = none;
+    util::Optional<std::string> authorization_header_name;
     std::map<std::string, std::string> custom_http_headers;
 
     util::Optional<std::string> url_prefix = none;
 
     // The name of the directory which Realms should be backed up to following
     // a client reset
-    util::Optional<std::string> recovery_directory = none;
+    util::Optional<std::string> recovery_directory;
+    ClientResyncMode client_resync_mode = ClientResyncMode::Recover;
```

Comment:
The argument was that we're turning an error scenario into a non-error rather than changing the behavior of correct things. The assumption is that in normal usage when running against a perfect server with infinite resource a client reset error will never occur, so client resync just lets us pretend that we're in that scenario more often than we actually are.